File: saw.f90
program saw
    integer,parameter::NMAX=20
    integer n,grid(0:2*NMAX,0:2*NMAX)
    double precision z,rsq
    common /integers/ n,grid
    common /floats/ z,rsq
    do i=0,2*NMAX,1
        do j=0,2*NMAX,1
            grid(i,j)=0
        enddo
    enddo
    print*,'#n[1];z[2];<rsq>[3]'
    do n=0,NMAX-1,1
        z=0.0
        rsq=0.0
        call mov(NMAX,NMAX,0)
        print*,n,';',z,';',rsq/z
    enddo
end

recursive subroutine mov(x,y,ln)
    integer,intent(in)::x
    integer,intent(in)::y
    integer,intent(in)::ln
    integer,parameter::NMAX=20
    integer n,grid(0:2*NMAX,0:2*NMAX)
    double precision z,rsq
    common /integers/ n,grid
    common /floats/ z,rsq
    if(ln .eq. n) then
        z=z+1
        rsq=rsq+(x-NMAX)**2+(y-nmax)**2
    else
        grid(x,y)=1
        if(grid(x+1,y) .eq. 0) then
            call mov(x+1,y,ln+1)
        endif
        if(grid(x-1,y) .eq. 0) then
            call mov(x-1,y,ln+1)
        endif
        if(grid(x,y+1) .eq. 0) then
            call mov(x,y+1,ln+1)
        endif
        if(grid(x,y-1) .eq. 0) then
            call mov(x,y-1,ln+1)
        endif
        grid(x,y)=0
    endif
end

File: saw.r
#!/usr/bin/Rscript

NMAX=20
n=0
z=0
rsq=0.0
grd=array(FALSE,dim=c(2*NMAX+1,2*NMAX+1))

printf=function(...) cat(sprintf(...))

mov=function(x,y,ln)
{
    if(ln==n)
    {
        z<<-z+1
        rsq<<-rsq+(x-NMAX-1)^2+(y-NMAX-1)^2
    }
    else
    {
        grd[x,y]<<-TRUE
        if(!grd[x+1,y])
        {
            mov(x+1,y,ln+1)
        }
        if(!grd[x-1,y])
        {
            mov(x-1,y,ln+1)
        }
        if(!grd[x,y+1])
        {
            mov(x,y+1,ln+1)
        }
        if(!grd[x,y-1])
        {
            mov(x,y-1,ln+1)
        }
        grd[x,y]<<-FALSE
    }
}

printf("#n[1];z[2];<rsq>[3]\n");
for(n in 0:(NMAX-1))
{
    z=0
    rsq=0.0
    mov(NMAX+1,NMAX+1,0)
    printf("%d;%.17g;%.17g\n",n,z,rsq/z)
}
